### ISSUE TITLE
Transform incorrectly calculates date bucket after updating older data

### ISSUE THREAD

**Issue by ritesh-vaswani** [elastic/elasticsearch#97101]
### Elasticsearch Version

8.6.2

### Java Version

_bundled_

### OS Version

Linux 20.04.2-Ubuntu

### Problem Description

Continuous tranform giving invalid result when date_histogram is used in pivot 

### Steps to Reproduce

// Step 1 Add ingest pipeline
PUT /_ingest/pipeline/es-timeadd_1?pretty
{
  "processors": [
    {
      "set": {
        "field": "_source.es_timestamp",
        "value": "{{_ingest.timestamp}}"
      }
    }
  ]
}

// Step2 Create index using above pipeline
PUT /test
{
  "settings": {
    "index.default_pipeline" : "es-timeadd"
  }
}

// Step3 Insert two documents
PUT /test/_doc/1
{
  "id" : "ritesh",
  "gen_timestamp_min" : "2023-06-21T15:54:00.000Z",
  "flow_count" : 5
}

PUT /test/_doc/2
{
  "id" : "ritesh",
  "gen_timestamp_min" : "2023-06-21T15:55:00.000Z",
  "flow_count" : 5
}

// Step4 - Create transform
PUT /_transform/test-transform
{
  "source": {
    "index": "test"
  },
  "dest" : {
    "index" : "test-transformed-index"
  },
  "sync": {
    "time": {
      "field": "es_timestamp",
      "delay": "0s"
    }
  },
  "pivot": {
    "group_by": {
      "gen_timestamp_hour": {
        "date_histogram": {
          "field": "gen_timestamp_min",
          "fixed_interval": "1h"
        }
      },
      "asset": {
        "terms": {
          "field": "id.keyword"
        }
      }
    },
    "aggregations": {
      "flow_count": {
        "sum": {
          "field": "flow_count"
        }
      }
    }
  }
}

// Step 5 - Start Transform
POST /_transform/test-transform/_start

// Wait for few seconds
// Step6 - Check data in destination index
GET test-transformed-index/_search
{
}

We get below output, In which sum of flow counts is 10 which is valid
{
  "took": 0,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 1,
      "relation": "eq"
    },
    "max_score": 1,
    "hits": [
      {
        "_index": "test-transformed-index",
        "_id": "cgAnWNlrvx_jB6WvSZa9f0n5AAAAAAAA",
        "_score": 1,
        "_source": {
          "gen_timestamp_hour": "2023-06-21T15:00:00.000Z",
          "flow_countNAME": 10,
          "asset": "ritesh"
        }
      }
    ]
  }
}

// Step 7 Update document in source index with id 1, as below
PUT /test/_doc/1
{
  "id" : "ritesh",
  "gen_timestamp_min" : "2023-06-21T15:54:00.000Z",
  "flow_count" : 6
}

// Step 8 Check data in output index, Total flow count should be 11, instead we see only 6 as below
GET test-transformed-index/_search
{
  
}

{
  "took" : 2,
  "timed_out" : false,
  "_shards" : {
    "total" : 1,
    "successful" : 1,
    "skipped" : 0,
    "failed" : 0
  },
  "hits" : {
    "total" : {
      "value" : 1,
      "relation" : "eq"
    },
    "max_score" : 1.0,
    "hits" : [
      {
        "_index" : "test-transformed-index",
        "_type" : "_doc",
        "_id" : "cgAnWNlrvx_jB6WvSZa9f0n5AAAAAAAA",
        "_score" : 1.0,
        "_source" : {
          "gen_timestamp_hour" : "2023-06-21T15:00:00.000Z",
          "flow_countNAME" : 6.0,
          "asset" : "ritesh"
        }
      }
    ]
  }
}

Root Cause - Which reprocessing in continuous tranform, It picks data upto gen_timestamp_min of modified document, It should process all the data in given group by.

**Comment by hendrikmuhs** [comment]
From you config:

```
"sync": {
  "time": {
    "field": "es_timestamp",
    "delay": "0s"
}
```

`delay` can never be `0s`. You created index uses defaults, Lucene's default refresh interval is `1s`.

Please try again with a reasonable value for `delay`, e.g. `2s`.

(It is still a bug that we allow setting `delay` to `0s`, if the above fixes your problem, feel free to change the bug report accordingly)

**Comment by ritesh-vaswani** [comment]
It is reproducible even with 30s delay

**Comment by przemekwitek** [comment]
> Root Cause - Which reprocessing in continuous tranform, It picks data upto gen_timestamp_min of modified document, It should process all the data in given group by.

I think you're right that the problematic part here is updating the record from the past which is then not reflected in the destination index.
This issue definitely requires further investigation.

**Comment by divyansh-pratap** [comment]
@hendrikmuhs can I pick this up as my first task.

**Comment by shreedaddy** [comment]
@hendrikmuhs @kingherc I am just getting started today as contributor. Please let me know if I can pickup this issue. Also appreciate if you can point me in the right to get started on this project.

**Comment by shreedaddy** [comment]
*in the right direction

**Comment by kingherc** [comment]
I'll leave this to @hendrikmuhs to tell you who's the more experienced with the ml part.

**Comment by hendrikmuhs** [comment]
@divyansh-pratap @shreedaddy /CC @kingherc 

I am happy if one of you picks this up. The fix is quite easy, I already provided it in https://github.com/elastic/elasticsearch/issues/97101#issuecomment-1611273486

```
            if (lowerBoundResult != null && upperBoundResult != null) {
                // we only need to round the lower bound, because the checkpoint will not contain new data for the upper bound
                lowerBound = rounding.round((long) lowerBoundResult.value());
                upperBound = (long) upperBoundResult.value();

                return false;
            }
```

I think `upperBound` must be rounded up:

```
                upperBound = rounding.nextRoundingValue(rounding.round((long) upperBoundResult.value());
```

However, that's the _easy_ part!

The fix _must_ come together with tests, I am expecting:

- unit test(s) in `CompositeBucketsChangeCollectorTests`
- an integration test, this _could_ be part of `TransformPivotRestIT`, there are some tests with "Continuous" in the name, but I am open for any other approach that fits

Feel free to ask any questions, I will try my best to answer them.

**Comment by divyansh-pratap** [comment]
Sure., Iwill pick this up and will also update the UT's and IT's.

**Comment by shreedaddy** [comment]
As this my first day on this open source project can I please get some guidance on how to get started like setting up the project if it is possible to run the project locally etc.
ThanksSree
    On Tuesday, July 11, 2023 at 07:26:11 PM GMT+5:30, Hendrik Muhs ***@***.***> wrote:  
 
 


@divyansh-pratap @shreedaddy /CC @kingherc

I am happy if one of you picks this up. The fix is quite easy, I already provided it in #97101 (comment)
            if (lowerBoundResult != null && upperBoundResult != null) {
                // we only need to round the lower bound, because the checkpoint will not contain new data for the upper bound
                lowerBound = rounding.round((long) lowerBoundResult.value());
                upperBound = (long) upperBoundResult.value();

                return false;
            }

I think upperBound must be rounded up:
                upperBound = rounding.nextRoundingValue(rounding.round((long) upperBoundResult.value());

However, that's the easy part!

The fix must come together with tests, I am expecting:
   
   - unit test(s) in CompositeBucketsChangeCollectorTests
   - an integration test, this could be part of TransformPivotRestIT, there are some tests with "Continuous" in the name, but I am open for any other approach that fits

Feel free to ask any questions, I will try my best to answer them.

**Comment by hendrikmuhs** [comment]
Please have a look at our contributing guidelines: https://github.com/elastic/elasticsearch/blob/main/CONTRIBUTING.md

**Comment by alyokaz** [comment]
I've refactored the unit tests for `DateHistogramFieldCollector`, splitting it into smaller tests and moving to a sepearte file, and wondered if this would, or should,  have to be part of a pull request for the issue or not? I'm aware, as stated in the `CONTRIBUTING.md`,  small, isolated refactorings are generally not considered worth the review effort and are rejected, but also fixes and refactoring shouldn't generally be combined.

**Comment by hendrikmuhs** [comment]
@alyokaz Fixes and refactorings in 1 PR is ok if you use separate commits for them. The re-factoring commit should be 1st, the fix and enhancements after. That way the reviewer can review the the fix separately from the refactoring.

**Comment by alyokaz** [comment]
@hendrikmuhs Understood. Thanks.

**Comment by imran-arshad** [comment]
Has this been fixed? I am facing the same issue on `v8.9.1`. In the meantime, is there a workaround for this other than downgrading the version?

**Comment by droberts195** [comment]
It’s fixed by https://github.com/elastic/elasticsearch/pull/97992
